packages/platform-apple-helpers/src/lib/commands/sign/provisioningProfile.ts

```diff
@@ -72,3 +87,142 @@ export function extractCertificateName(subject: string) {
+  const provisioningEntitlements = await readKeyFromPlist(
+    provisioningPlistPath,
+    'Entitlements',
+    { xml: true },
+  );
+
+  // Extract entitlements from app binary
+  const appEntitlements = await extractAppEntitlements(appPath);
+
+  // Merge entitlements
+  const mergedEntitlements = await mergeEntitlements(
+    provisioningEntitlements,
+    appEntitlements,
+    provisioningPlistPath,
+  );
+
+  fs.writeFileSync(outputPath, mergedEntitlements);
+  logger.debug(`Generated merged entitlements file: ${relativeToCwd(outputPath)}`);
+}
+
+/**
+ * Extract entitlements from app binary using codesign.
+ */
+async function extractAppEntitlements(appPath: string): Promise<string> {
+  try {
+    const result = await spawn('codesign', ['-d', '--entitlements', '-', '--xml', appPath]);
+    return result.stdout.trim();
+  } catch (error) {
+    logger.debug(`Could not extract entitlements from app: ${(error as SubprocessError).stderr}`);
+    // Return empty plist if no entitlements found
+    return '<?xml version="1.0" encoding="UTF-8"?>\n<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">\n<plist version="1.0">\n<dict>\n</dict>\n</plist>';
+  }
+}
+
+/**
+ * Merge entitlements from app and provisioning profile.
+ * Based on fastlane's entitlements merging logic.
+ */
+async function mergeEntitlements(
+  provisioningEntitlements: string,
+  appEntitlements: string,
+  provisioningPlistPath: string,
+): Promise<string> {
+  // Write temporary files for processing
+  const tempDir = fs.mkdtempSync(path.join(process.cwd(), '.rock-temp-'));
+  const profileEntitlementsPath = path.join(tempDir, 'profile_entitlements.plist');
+  const appEntitlementsPath = path.join(tempDir, 'app_entitlements.plist');
+  const mergedEntitlementsPath = path.join(tempDir, 'merged_entitlements.plist');
+
+  try {
+    fs.writeFileSync(profileEntitlementsPath, provisioningEntitlements);
+    fs.writeFileSync(appEntitlementsPath, appEntitlements);
+
+    // Get team identifier from provisioning profile
+    const teamIdentifier = await readKeyFromPlist(provisioningPlistPath, 'TeamIdentifier:0');
+
+    // Define entitlements that should be transferred from app to profile
+    const transferRules = [
+      'com.apple.developer.icloud-container-identifiers',
+      'com.apple.developer.icloud-services',
+      'com.apple.developer.ubiquity-kvstore-identifier',
+      'com.apple.developer.icloud-container-environment',
+      'com.apple.security.application-groups',
+      'keychain-access-groups',
+      'com.apple.developer.associated-domains',
+      'com.apple.developer.healthkit',
+      'com.apple.developer.homekit',
+      'inter-app-audio',
+      'com.apple.developer.networking.networkextension',
+      'com.apple.developer.maps',
+      'com.apple.external-accessory.wireless-configuration',
+      'com.apple.developer.siri',
+      'com.apple.developer.nfc.readersession.formats',
+    ];
```

Comment:
Do we need to maintain this list of entitlements? Isn't there a way to avoid that? If not, where can we get the list from, so we can be sure it's in sync

Reply:
awesome, let's add it as a comment please